libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/helpers/DataClass.kt

```diff
@@ -62,17 +67,186 @@ fun TypeSpec.Builder.makeDataClassFromProperties(properties: List<PropertySpec>)
+        .addStatement("return true")
+        .endControlFlow()
+        .beginControlFlow("if (other is %T)", className)
+        .apply {
+          if (propertySpecs.isEmpty()) {
+            add("return true\n")
+          } else {
+            add("return %L", propertySpecs
+                .excludeInternalProperties()
+                .map { equalsCode(it) }
+                .joinToCode("&& ")
+            )
+          }
+        }
+        .endControlFlow()
+        .addStatement("return false")
+        .build()
+  }
+
+  return addFunction(FunSpec.builder(Identifier.equals)
+      .addModifiers(KModifier.OVERRIDE)
+      .addParameter("other", KotlinSymbols.Any.copy(nullable = true))
+      .returns(KotlinSymbols.Boolean)
+      .addCode(methodCode())
+      .build())
+}
+
+internal fun TypeSpec.Builder.withHashCodeImplementation(): TypeSpec.Builder = apply {
+  fun hashPropertyCode(property: PropertySpec) =
+      CodeBlock.builder()
+          .addStatement("${Identifier.__h} *= 31")
+          .apply {
+            if (property.type == typeNameOf<Boolean>()) {
+             addStatement("${Identifier.__h} += if (%L) 1 else 0", property.name)
+            } else {
+              addStatement("${Identifier.__h} += %L.hashCode()", property.name)
+            }
+          }
+          .build()
+
+  fun methodCode(): CodeBlock {
+    if (propertySpecs.isEmpty()) {
+      return CodeBlock.builder().addStatement("return·this::class.hashCode()").build()
+    }
+    return CodeBlock.builder()
+        .beginControlFlow("if (%L == null)", MEMOIZED_HASH_CODE_VAR)
+        .addStatement("var ${Identifier.__h} = %L.hashCode()", propertySpecs.getOrNull(0)?.name ?: "null")
+        .add(propertySpecs
+            .drop(1)
+            .excludeInternalProperties()
+            .map(::hashPropertyCode)
+            .fold(CodeBlock.builder(), CodeBlock.Builder::add)
+            .build())
+        .addStatement("%L = ${Identifier.__h}", MEMOIZED_HASH_CODE_VAR)
+        .endControlFlow()
+        .addStatement("return %L!!", MEMOIZED_HASH_CODE_VAR)
+        .build()
+  }
+
+  check(propertySpecs.all { !it.mutable }) {
+    "Cannot generate hashCode function for class with mutable properties, memoization would fail"
+  }
+
+  if (propertySpecs.isNotEmpty()) {
+    addProperty(
+        PropertySpec.builder(MEMOIZED_HASH_CODE_VAR, KotlinSymbols.Int.copy(nullable = true), KModifier.PRIVATE)
+            .mutable()
+            .initializer("null")
+            .build()
+    )
+  }
+
+  addFunction(FunSpec.builder(Identifier.hashCode)
+      .addModifiers(KModifier.OVERRIDE)
+      .returns(KotlinSymbols.Int)
+      .addCode(methodCode())
+      .build()
+  )
+}
+
+internal fun TypeSpec.Builder.withToStringImplementation(className: ClassName): TypeSpec.Builder {
+  fun printPropertiesTemplate() =
+      "${className.simpleName}(" + propertySpecs
+            .excludeInternalProperties()
+            .joinToString(",") { "${it.name}=\$${it.name}" } + ")"
```

Comment:
Adding a space makes the toString match the Kotlin version
```suggestion
            .joinToString(", ") { "${it.name}=\$${it.name}" } + ")"
```